### ISSUE TITLE
2025年12月31日，当前版本，多张显卡是不能选择某一显卡运行代码。

### ISSUE THREAD

**Issue by zyc01** [ultralytics/ultralytics#23089]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

2025年12月31日，当前版本，多张显卡是不能选择某一显卡运行代码。当使用一个月前的版本，将ultralytics文件夹中的models文件夹替换到当前版本，问题就会消失。这是新版本的bug吗？

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zyc01, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

🤖 This is an automated response to help us triage quickly—an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In your case (multi-GPU 环境下无法选择某一张显卡 + 替换 `models/` 文件夹后恢复)，MRE 里请尽量包含以下信息，便于我们复现定位：
-你使用的 `ultralytics` 具体版本号（当前版本与一个月前版本）
-操作系统、Python 版本、PyTorch 版本、CUDA 版本
-你希望选择的 GPU 编号、机器上实际 GPU 数量
-你运行的最小代码/CLI 命令（只保留能复现问题的最少行）
-运行日志与报错信息（如有）
-你提到的替换行为：替换的是哪个版本的 `models/`，替换后具体现象变化是什么

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
What's the code you're running?

**Comment by wangyutong-bit** [comment]
可以呀，用Liunx系统指定某张卡，像我有四张卡，用[0,1]或者说[2,3]都可以训练呀，对了，你不会说的是运行吧，我用的liunx系统，指定GPU也可以运行呀

**Comment by glenn-jocher** [comment]
@wangyutong-bit 说得对，当前版本正常情况下可以通过 `device` 显式指定单卡/多卡；@zyc01 这类“指定不了某张卡”的情况更常见是环境里有多重安装/路径冲突或进程启动前 CUDA 默认初始化导致的，并不一定是新版本 bug（请先不要通过替换 `ultralytics/models` 目录来混用版本）。参考 [Train 模式多 GPU 与设备选择说明](https://docs.ultralytics.com/modes/train/)。

麻烦贴一下你实际运行的命令/代码与完整日志（包含打印出的 `device` 行），并把下面两段输出发出来方便定位：

```bash
yolo checks
```

```python
from ultralytics.utils.checks import collect_system_info
print(collect_system_info())
```

如果你说的“运行”是推理/预测，也请显式传 `device=2`（或在启动进程前设置 `CUDA_VISIBLE_DEVICES=2`），避免被默认 GPU 抢占。

**Comment by Farenweh** [comment]
出现这种问题的常见原因是你在Ultralytics框架内部启动ddp子线程之前初始化了cuda上下文，常见的操作是使用了cuda.is_available等，可能是引入某些依赖于torch cuda的第三方库或手动检测是否为cuda环境导致的。

**Comment by glenn-jocher** [comment]
@Farenweh 这个方向基本对：如果在启动 DDP/多进程之前任何地方触发了 CUDA 上下文（例如 `torch.cuda.is_available()`、`torch.cuda.get_device_properties()`，或某些第三方库导入时隐式调用 CUDA），进程往往会先在默认 `CUDA:0` 初始化，后面再传 `device=` 或改 `CUDA_VISIBLE_DEVICES` 就可能不再生效（因为 CUDA 已经初始化了）；Ultralytics 的设备选择逻辑也依赖在“CUDA 未初始化前”生效，细节可参考 [`select_device()` 说明](https://docs.ultralytics.com/reference/utils/torch_utils/)。  

建议你把 GPU 选择放到脚本最顶部、并在 `import torch/ultralytics` 之前设置环境变量，且把所有 CUDA 探测/初始化代码移到 `if __name__ == "__main__":` 之后，例如：  
```python
import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "2"  # 只暴露物理第2号卡给本进程

from ultralytics import YOLO

if __name__ == "__main__":
    model = YOLO("yolo11n.pt")
    model.train(data="coco8.yaml", device=0)  # 这里的0表示“可见设备里的第0张”，即物理GPU 2
```  
如果你方便的话，请贴一份“最小可复现脚本”（尤其是训练/推理前有没有任何 `torch.cuda.*` 调用或相关库导入），我们就能更快判断是环境初始化顺序问题还是确实存在回归。
